### PR TITLE
[FIX] html_editor: trailing br at the end of link

### PR DESCRIPTION
**Pull request by adsa-odoo**
**Current behaviour before PR:**

Creating a link in an empty block using powerbox contains trailing `br` at the end of link. Due to this `br` , when user presses `ctrl + A` and applies any format, `br` also gets formatted along with link. In such case when user selects link using double click and removes format using remove-format button, the link gets unformatted but `br` remains formatted because it is not traversed in selection.

**Desired behaviour after PR:**

There should be no trailing `br` after creating a link as the block is not empty anymore.

task-4399010

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/html_editor/static/src/main/link/link_plugin.js [hunk 1]
```diff
@@ -397,6 +398,7 @@ export class LinkPlugin extends Plugin {
                     } else {
                         this.linkElement.removeAttribute("class");
                     }
+                    cleanTrailingBR(closestBlock(this.linkElement));
```
Comment (deso-odoo) [inline addons/html_editor/static/src/main/link/link_plugin.js:401]: I agree here that by removing the `br` here would eliminate the issue of applying styles. But what about the linebreak `br`. Let's say you have `<p><a>[abc.com</a><br>]<br></p>` => apply font size => `<p><a><span class="someFontSizeClass">[abc.com</span></a><span class="someFontSizeClass"><br></span>]<br></p>` => now select the text => `<p><a><span class="someFontSizeClass">[abc.com]</span></a><span class="someFontSizeClass"><br></span><br></p>` => removeFormat => and it will result in `<p><a>[abc.com]</a><span class="someFontSize"><br></span><br></p>` and the bug will still occur. I think this will also need a fix in `formatSelection` where it should filter out the linebreak `br`, specifically those representing line breaks, not the ones indicating empty lines.

Comment (adsa-odoo) [reply addons/html_editor/static/src/main/link/link_plugin.js:401]: The case you mentioned should indeed not format the `<br>` but since [this](https://github.com/odoo/odoo/commit/df1dcf1c6bfc94ae8e084b41c1b5cc84b2cc5203) commit has merged, all the selected `<br>`s will get formatted. Please note that we can't filter out the line-break `<br>`s and format only the ones indicating empty lines because we will have problem if we are dealing with multiple line breaks: `<p>test[<br><br><br><br><br>]</p>` .
Right now it works fine by wrapping all the `br`s in a single `span` . However, if the commit I mentioned only addresses the issue related to table (as mentioned in the commit message) then we can fix that commit and the format will not be applied to any `br` (as it was before).

Comment (deso-odoo) [reply addons/html_editor/static/src/main/link/link_plugin.js:404]: :+1: 
Can you add a test case for the case I mentioned. :slightly_smiling_face:

**Comment by dmo-odoo** [comment]
Hello, this PR is in conflict. Also, there is only one test so I'm not sure if it's the test that @deso-odoo asked to have added or not.

**Comment by adsa-odoo** [comment]
Hello @dmo-odoo, one test case is written in first commit and 2nd test case is written in 2nd commit which is the one asked by @deso-odoo .

File: addons/html_editor/static/src/core/format_plugin.js [hunk 2]
```diff
@@ -236,7 +238,9 @@ export class FormatPlugin extends Plugin {
                 .filter(
                     (n) =>
                         ((isTextNode(n) && (isVisibleTextNode(n) || isZWS(n))) ||
-                            n.nodeName === "BR") &&
+                            (n.nodeName === "BR" &&
+                                (isFakeLineBreak(n) ||
+                                    previousLeaf(n, closestBlock(n))?.nodeName === "BR"))) &&
```
Comment (dmo-odoo) [inline addons/html_editor/static/src/core/format_plugin.js:243]: Shouldn't `isFakeLineBreak` return `true` in the case of `previousLeaf(n, closestBlock(n))?.nodeName === "BR"))` ? I don't understand what case it covers. I suppose it has something to do with the test below but even then it's hard to tell without investigating. Is it because the BR is in a span or something?

Comment (adsa-odoo) [reply addons/html_editor/static/src/core/format_plugin.js:243]: Here second condition is useful to cover the case if there are more than two `<br>`s. Let's assume this case: 
`<p>abc<br><br><br></p>`
Here, for 1st `<br>` `isFakeLineBreak(n)` is false. In this case `previousLeaf(n, closestBlock(n))?.nodeName === "BR")` is false because we don't want the 1st `<br>`.

For 2nd `<br>`, isFakeLineBreak(n) is false, but we want to include that `<br>`, that's why `previousLeaf(n, closestBlock(n))?.nodeName === "BR")` is true and 2nd `<br>` will be included too.

For 3rd `<br>`, isFakeLineBreak(n) is true since it is the last one, we want to include 3rd `<br>` without caring 2nd condition.
In other words, 1st `<br>` is not included and rest of them are included.

Comment (dmo-odoo) [reply addons/html_editor/static/src/core/format_plugin.js:243]: I see, but if the code only makes sense for more than 2 BR then the test should reflect that and have more than two BRs. Maybe multiple tests for 2 and 3 or 4 BRs?

Comment (adsa-odoo) [reply addons/html_editor/static/src/core/format_plugin.js:243]: I have added one more test case for 3 `<br>`.

**Comment by dmo-odoo** [comment]
@adsa-odoo Thanks!
@robodoo rebase-ff r+
